apps/grader-host/src/index.js

```diff
@@ -457,311 +384,238 @@ function runJob(info, callback) {
+        IpcMode: 'private',
+        CpuPeriod: config.graderDockerCpuPeriod,
+        CpuQuota: config.graderDockerCpuQuota,
+        PidsLimit: config.graderDockerPidsLimit,
+        Ulimits: [
           {
-            stream: true,
-            stdout: true,
-            stderr: true,
-          },
-          (err, stream) => {
-            if (ERR(err, callback)) return;
-            const out = byline(stream);
-            out.on('data', (line) => {
-              logger.info(`container> ${line.toString('utf8')}`);
-            });
-            callback(null, container);
+            // Disable core dumps, which can get very large and bloat our storage.
+            Name: 'core',
+            Soft: 0,
+            Hard: 0,
           },
-        );
-      },
-      (container, callback) => {
-        container.start((err) => {
-          if (ERR(err, callback)) return;
-          logger.info('Container started!');
-          callback(null, container);
-        });
+        ],
       },
-      async (container) => {
-        results.start_time = await timeReporter.reportStartTime(jobId);
-        return container;
-      },
-      async (container) => {
-        const timeoutId = setTimeout(() => {
-          results.timedOut = true;
-          container.kill().catch((err) => {
-            globalLogger.error('Error killing container', err);
-          });
-        }, runTimeout * 1000);
+      Entrypoint: entrypoint.split(' '),
+    });
 
-        logger.info('Waiting for container to complete');
-        try {
-          await container.wait();
-        } finally {
-          clearTimeout(timeoutId);
-        }
+    const stream = await container.attach({
+      stream: true,
+      stdout: true,
+      stderr: true,
+    });
+    const out = byline(stream);
+    out.on('data', (line) => {
+      logger.info(`container> ${line.toString('utf8')}`);
+    });
 
-        return container;
-      },
-      async (container) => {
-        results.end_time = await timeReporter.reportEndTime(jobId);
-        return container;
-      },
-      (container, callback) => {
-        container.inspect((err, data) => {
-          if (ERR(err, callback)) return;
-          if (results.timedOut) {
-            logger.info('Container timed out');
-          } else {
-            logger.info(`Container exited with exit code ${data.State.ExitCode}`);
-          }
-          results.succeeded = !results.timedOut && data.State.ExitCode === 0;
-          callback(null, container);
-        });
-      },
-      (container, callback) => {
-        container.remove(
-          {
-            // Remove any volumes associated with this container
-            v: true,
-          },
-          (err) => {
-            if (ERR(err, callback)) return;
-            callback(null);
-          },
-        );
-      },
-      (callback) => {
-        // We made it through the Docker danger zone!
-        clearTimeout(jobTimeoutId ?? undefined);
-        jobTimeoutId = null;
```

Comment:
Calling this out for review: AFAICT, we don't actually need to null this out for the purposes of avoiding calling `clearTimeout` multiple times. Node will silently do nothing on subsequent calls on `clearTimeout` with the same timer object.